### PR TITLE
Add async version of the StartProcessingAsync for event streams

### PR DESCRIPTION
**Pull request by normj**
## Description
https://github.com/aws/aws-sdk-net/issues/3360

When users use response streaming for APIs like Bedrock the `StartProcessing` method eventually goes down to a blocking read on the network stream. This uses up a thread for the process which can cause scaling concerns.

This PR creates an async version of the `StartProcessing` method all the way down to the network read. This new method has the benefit that callers can await the `Task` returned from `StartProcessingAsync` which pausing execution in the thread that called `StartProcessingAsync` till all of the events are processed.

```csharp
var invokeResponse = await bedrockClient.InvokeModelWithResponseStreamAsync(invokeRequest);

invokeResponse.Body.ChunkReceived += Body_ChunkReceived;
await invokeResponse.Body.StartProcessingAsync();

Console.WriteLine("Done");

async void Body_ChunkReceived(object? sender, Amazon.Runtime.EventStreams.EventStreamEventReceivedArgs<PayloadPart> e)
{
    Console.WriteLine(UTF8Encoding.UTF8.GetString(e.EventStreamEvent.Bytes.ToArray()));
}
```

## Testing
Added new integ tests and wrote test app to confirm behavior.

Internal dry run was successful.

### CODE REVIEW

**Comment by danielmarbach** [comment]
@normj something is off with this entire API. StartProcessing like the name suggests starts back processing which the when data is available invokes the corresponding event handlers. So the method is designed to be fire and forgot so that the caller can continue and effectively it is the responsibility of the caller to have synchronization primitives in place to know when the event handlers are called. This PR seems to deviate from the design to solve the thread blocking problem but introduces weird assymetry on the API surface. Technical rhe existing start method can use the async path already which would remove the thread offloading and unlock the threads during the IO bound operation without having to introduce a new API

If you look at what the user requests that would also address the threading part. Then the next problem is that the higher level event APIs are not async friendly which an async enumerable would solve

**Comment by normj** [comment]
> @normj something is off with this entire API. StartProcessing like the name suggests starts back processing which the when data is available invokes the corresponding event handlers. So the method is designed to be fire and forgot so that the caller can continue and effectively it is the responsibility of the caller to have synchronization primitives in place to know when the event handlers are called. This PR seems to deviate from the design to solve the thread blocking problem but introduces weird assymetry on the API surface. Technical rhe existing start method can use the async path already which would remove the thread offloading and unlock the threads during the IO bound operation without having to introduce a new API
> 
> If you look at what the user requests that would also address the threading part. Then the next problem is that the higher level event APIs are not async friendly which an async enumerable would solve

I also updated the original `StartProcess` method to use the new async code path down to the network stream getting rid of the blocking stream. Except for .NET Framework 3.5. So user's request would be satisfied by updating to the SDK.

```csharp
        public virtual void StartProcessing()
        {
            if (IsProcessing) return;

            IsProcessing = true;
            Process();
        }

        protected void Process()
        {
#if AWS_ASYNC_API
            // Task only exists in framework 4.5 and up, and Standard.
            _ = ProcessLoopAsync();
#else
            // ThreadPool only exists in 3.5 and below. These implementations do not have the Task library.
            ThreadPool.QueueUserWorkItem(ProcessLoop);
#endif
        }
```

I thought it would be useful to add the `StartProcessingAsync` returning the `Task` because like you said "responsibility of the caller to have synchronization primitives in place to know when the event handlers are called." but the SDK doesn't give any primitives back to the caller to use. And there is no telling how many times the callbacks will be called. So I thought if I gave back the `Task` the user could choose when to end for example the HTTP response from a server while using the callbacks to send data through the response while data is coming back from Bedrock. If they don't care they can always ignore the `Task` and use it for fire and forget.

In the PR I created an async version of the existing S3 select integ test. You can see in that test when using the feature it was written waiting till it got the known number of callbacks or a timeout. That works if you know the number of callbacks you should get but likely you don't know especially in Bedrocks case.

```csharp
[TestMethod]
[TestCategory("S3")]
public void TestCallEvents()
{
    var eventStream = GetSelectObjectContentEventStream();

    var testContent = "";
    var recordsEvents = new List<RecordsEvent>();

    eventStream.RecordsEventReceived += (sender, args) => recordsEvents.Add(args.EventStreamEvent);
    eventStream.StartProcessing();
    SpinWait.SpinUntil(() => recordsEvents.Count > 0, TimeSpan.FromSeconds(5));

    using (var streamReader = new StreamReader(recordsEvents[0].Payload))
    {
        testContent = streamReader.ReadToEnd();
    }
    eventStream.Dispose();

    AssertRecordsEqualsExpected(testContent);
}
```

The event stream API is not my favorite part of the SDK's user experience. The original request asked for the use of `IAsyncEnumerable` user experience which I agree would be better for most use cases. But that is a larger effort that I have bandwidth right now and this seemed like a reasonable way to unblock the original request. The event stream API predates `IAsyncEnumerable` existing which is why it wasn't done that way from the start.

File: sdk/src/Core/Amazon.Runtime/EventStreams/Internal/EventStream.cs
Comment (danielmarbach) [inline sdk/src/Core/Amazon.Runtime/EventStreams/Internal/EventStream.cs:265]: You still want to wrap this async version with a Task.Run. Otherwise, the caller is still blocked until the first actual IO bound operation happens, which is deep down in the network stream read.

**Comment by danielmarbach** [comment]
> The event stream API is not my favorite part of the SDK's user experience. The original request asked for the use of `IAsyncEnumerable` user experience which I agree would be better for most use cases. But that is a larger effort that I have bandwidth right now and this seemed like a reasonable way to unblock the original request.

I think my comment came primarily from the fact that both methods are called `StartProcessing`, but they behave differently. One starts the processing by starting to call the event handler on a background thread and leaves it up to the caller to do the necessary synchronization while the other starts the processing but then holds up the calling thread until the processing is done. This is confusing to me from an API surface perspective.

I would argue either you have `StartProcessing` starting to process and raising the event and then a `StopProcessing` and `StopProcessingAsync` that allows the caller to either synchronously wait or asynchronously "wait" until the processing is done, or you would have to rename the new async method to be something like `StartProcessingAndWaitForCompletionAsync`.

File: sdk/src/Core/Amazon.Runtime/EventStreams/Internal/EventStream.cs
Comment (peterrsongg) [inline sdk/src/Core/Amazon.Runtime/EventStreams/Internal/EventStream.cs:275]: How do you feel about optimizing here for memory? Something like
```
#If !NETFRAMEWORK
            var arrayPool = ArrayPool<byte>.Shared;
            var buffer = arrayPool.Rent(BufferSize);
#else
             var buffer = new byte[BufferSize];
#endif
```
Then conditionally adding the using statement based off `!NETFRAMEWORK`

Comment (danielmarbach) [reply sdk/src/Core/Amazon.Runtime/EventStreams/Internal/EventStream.cs:275]: FYI I would be very careful with pooling here. The problem is that the byte array is handed out to an event handler that then theoretically can do whatever with the byte array. While this problem already exists today with the current approach at least the byte array is local to the EventStream implementation and therefore "buffered" locally over subsequent reads until no data is available anymore. With the pool you would acquire a buffer from the pool and keep it for a long period and then return the buffer to the pool. If the event handler does something weird that buffer reference would be handed back to the pool and then eventually handed out to someone else

If you want to do pooling it might be better to pool for every read and return after the handler returned or does the decoder do some memory copying under the covers?

Comment (peterrsongg) [reply sdk/src/Core/Amazon.Runtime/EventStreams/Internal/EventStream.cs:275]: that's a good point, the decoder does do some memory copying under the covers, but the added complexity of pooling probably isn't worth the benefit here.

**Review by peterrsongg — APPROVED** [review]

**Comment by danielmarbach** [comment]
> > The event stream API is not my favorite part of the SDK's user experience. The original request asked for the use of `IAsyncEnumerable` user experience which I agree would be better for most use cases. But that is a larger effort that I have bandwidth right now and this seemed like a reasonable way to unblock the original request.
> 
> I think my comment came primarily from the fact that both methods are called `StartProcessing`, but they behave differently. One starts the processing by starting to call the event handler on a background thread and leaves it up to the caller to do the necessary synchronization while the other starts the processing but then holds up the caller until the processing is done. This is confusing to me from an API surface perspective.
> 
> I would argue either you have `StartProcessing` starting to process and raising the event and then a `StopProcessing` and `StopProcessingAsync` that allows the caller to either synchronously wait or asynchronously "wait" until the processing is done, or you would have to rename the new async method to be something like `StartProcessingAndWaitForCompletionAsync`.

@peterrsongg what do you think about the comment above?

File: sdk/src/Core/Amazon.Runtime/EventStreams/Internal/EventStream.cs
Comment (normj) [reply sdk/src/Core/Amazon.Runtime/EventStreams/Internal/EventStream.cs]: Good call on adding the `Task.Run`. Added that back in.

**Comment by normj** [comment]
> starts the processing but then holds up the caller until the processing is done.

I don't view that we are forcing the caller to be held up because the caller can choose to await the `Task` or not. It gives the caller the option to use the `Task` as a synchronization context. I don't think a `StopProcessingAsync` makes sense because a caller doesn't have any knowledge how far they are in the data being sent through the stream and wouldn't have any context on when to stop. I can kind of see the argument for a renaming to `StartProcessingAndWaitAsync` but the reality the method isn't the one waiting. It is up to the caller to choose to wait by awaiting the `Task` returned.

**Comment by danielmarbach** [comment]
I brought up the term Stop because that is what message / event pumps usually use to indicate as stop processing and wait until currently processed items are successfully processed but I do understand your concern there that Stop is not the right terminology. 

I'm mostly concerned about the API behavioral differences. When you look how other message or event processors work is the essentially represent a pump. When you call `Start` you start the pump. Any task returned there would represent the IO-bound operations that may occur until the pump is started.

The synchronous start method exactly works like that. The async version would behave differently. So from an API perspective you can argue, even in the synchronous case when you have a way to start the pump you also want to have a way to indicate that you would like to wait until the pump is done. So by splitting the starting from "waiting for done or all signals arrived" into dedicated API methods you would get that, and it would be symmetrical across the synchronous and asynchronous code path, which, I think, is ultimately the more consistent behavior from the SDK.

**Comment by peterrsongg** [comment]
> would be symmetrical ac

So is your concern that `StartProcessingAsync()` contains both the logic for starting the pump and  waiting until all events are done processing and that you'd like them separated out? Also just so I understand correctly, are you saying that "waiting for done **or** all signals arrived" should be separate APIs? All signals arriving would signal that we are done, so this part has me a bit confused.

The original purpose of this PR was to get rid of the blocking IO network read, and adding this functionality would cause the scope of this PR to get bigger than intended, but perhaps this is something we could address in a separate PR.

**Comment by peterrsongg** [comment]
> I'm mostly concerned about the API behavioral differences. When you look how other message or event processors work is the essentially represent a pump. When you call `Start` you start the pump. Any task returned there would represent the IO-bound operations that may occur until the pump is started.
> 
> The synchronous start method exactly works like that. The async version would behave differently. So from an API perspective you can argue, even in the synchronous case when you have a way to start the pump you also want to have a way to indicate that you would like to wait until the pump is done. So by splitting the starting from "waiting for done or all signals arrived" into dedicated API methods you would get that, and it would be symmetrical across the synchronous and asynchronous code path, which, I think, is ultimately the more consistent behavior from the SDK.

So is your concern that `StartProcessingAsync()` contains both the logic for starting the pump and  waiting until all events are done processing and that you'd like them separated out? Also just so I understand correctly, are you saying that "waiting for done **or** all signals arrived" should be separate APIs? All signals arriving would signal that we are done, so this part has me a bit confused.

The original purpose of this PR was to get rid of the blocking IO network read, and adding this functionality would cause the scope of this PR to get bigger than intended, but perhaps this is something we could address in a separate PR.

**Comment by danielmarbach** [comment]
> The original purpose of this PR was to get rid of the blocking IO network read, and adding this functionality would cause the scope of this PR to get bigger than intended, but perhaps this is something we could address in a separate PR.

My input early on was that this could have been addressed by not adding the StartProcessingAsync method and simply within the StartProcessing method have two implementation paths. One that offloads to the worker thread pool and uses synchronous IO-bound operation to read until completed and one that does a Task.Run that calls into a path that only uses the asynchronous stream APIs like mentioned here

> This PR seems to deviate from the design to solve the thread blocking problem but introduces weird asymmetry on the API surface. Technical the existing start method can use the async path already which would remove the thread offloading and unlock the threads during the IO bound operation without having to introduce a new API

By adding the new StartProcessingAsync the scope of this PR was in my view already expanded and that's why I brought my thoughts up here. 

I realize though nobody has asked me about my opinion and from this point on I might come across like a broken record and I definitely don't want to be that guy.

**Comment by normj** [comment]
@danielmarbach your opinions are welcome, don't feel you need to be asked.

What I'm wrestling with is the `StartProcessing` is pretty challenging to use correctly for API's like Bedrock's event stream. I'm thinking in v4 we should deprecate `StartProcessing` in fact and could be convinced to do it in v3. The event stream mechanism was added when S3 added their [SelectObjectContent](https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html#API_SelectObjectContent_ResponseElements) operation. In that API it returns multiple events including an `EndEvent`. So callers could use the event stream API and run the operation till they got the `EndEvent`.

Later as Lambda and now Bedrock have added response streaming APIs they were modeled with no ending event. So the assumption used for `S3` that users could listen to events till there was an end event was invalidated. So I'm not sure how you would ever use `StartProcessing` without putting basically a sleep hoping it finishes with in that time. That approach has serious problems.

The reason I don't deprecate `StartProcessing` in v3 is it still has the .NET Framework support with no `Task` support. I agree that `StartProcessing` and `StartProcessingAsync` have different semantics but I would rather consider `StartProcessing` as obsolete and not be considered the driver for the behavior. Ultimately I expect most users will just need to await the `StartProcessingAsync` and asking them to know to call 2 different APIs every time seems like unnecessary work.

**Comment by boblodgett** [comment]
> @danielmarbach your opinions are welcome, don't feel you need to be asked.
> 
> What I'm wrestling with is the `StartProcessing` is pretty challenging to use correctly for API's like Bedrock's event stream. I'm thinking in v4 we should deprecate `StartProcessing` in fact and could be convinced to do it in v3. The event stream mechanism was added when S3 added their [SelectObjectContent](https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html#API_SelectObjectContent_ResponseElements) operation. In that API it returns multiple events including an `EndEvent`. So callers could use the event stream API and run the operation till they got the `EndEvent`.
> 
> Later as Lambda and now Bedrock have added response streaming APIs they were modeled with no ending event. So the assumption used for `S3` that users could listen to events till there was an end event was invalidated. So I'm not sure how you would ever use `StartProcessing` without putting basically a sleep hoping it finishes with in that time. That approach has serious problems.
> 
> The reason I don't deprecate `StartProcessing` in v3 is it still has the .NET Framework support with no `Task` support. I agree that `StartProcessing` and `StartProcessingAsync` have different semantics but I would rather consider `StartProcessing` as obsolete and not be considered the driver for the behavior. Ultimately I expect most users will just need to await the `StartProcessingAsync` and asking them to know to call 2 different APIs every time seems like unnecessary work.

I think this is reasonable and agree we need to make sure we obsolete `StartProcessing` as soon as possible without breaking customers. This sounds like a v4 change we need to record.

File: generator/.DevConfigs/E75E0238-4533-4EDC-89B5-08C4038EBE03.json
Comment (boblodgett) [inline generator/.DevConfigs/E75E0238-4533-4EDC-89B5-08C4038EBE03.json:3]: Non blocking. Change if making other changes: `apis use` => `APIs to use`

**Review by boblodgett — APPROVED** [review]

**Comment by lucasmeijer** [comment]
Good news: it works.

Bad news: it makes the thing people are most likely wanting to do very difficult.  I have to write this code myself in order to use bedrock.  This makes bedrock the odd one out, as my openAI, anthropic, azure, and groq implementations all do not ask me to write code like this.  Quite tricky code to get right, I'm not 100% sure I got it right.


```
static async IAsyncEnumerable<PayloadPart> PayloadPartsFrom(InvokeModelWithResponseStreamResponse r)
        {
            ConcurrentQueue<PayloadPart> queue = new();

            using SemaphoreSlim semaphore = new(0);
            
            void OnBodyOnChunkReceived(object? sender, EventStreamEventReceivedArgs<PayloadPart> args)
            {
                queue.Enqueue(args.EventStreamEvent);
                semaphore.Release();
            }
            
            r.Body.ChunkReceived += OnBodyOnChunkReceived; 
            var finishTask = r.Body.StartProcessingAsync();

            while (true)
            {
                var t = await Task.WhenAny(semaphore.WaitAsync(), finishTask);
                if (t == finishTask)
                    yield break;
                while (queue.TryDequeue(out var p))
                    yield return p;
            }
        }
```
 
That said, it does work, and allows me to move forward, so quite happy with that.

**Pull request merged**
